### PR TITLE
Small improvements to NJT matrix multiplies

### LINKED ISSUE #146404: Can't back prop through NJT matrix multiplication when output is strided tensor
### 🐛 Describe the bug

When performing a matmul between two NJTs such that the output is strided, it currently fails on the backward pass.
This is on latest nightly.

Repro:

```python
import torch

nt0 = torch.nested.nested_tensor([torch.rand(2, 6), torch.rand(3, 6)], layout=torch.jagged, requires_grad=True)
nt1 = torch.nested.nested_tensor_from_jagged(torch.rand(5, 6), offsets=nt0.offsets()).requires_grad_(True)

out = torch.matmul(nt0.transpose(-2, -1), nt1)
out.sum().backward()
```
Gives the following error:
```
NestedTensor matmul_backward_default(grad: jt_all, self: jt_all, other: any, mask: any): expected grad to be a jagged layout NestedTensor
```

This is because `out` is not a nested tensor, and the ops in https://github.com/pytorch/pytorch/blob/main/torch/nested/_internal/ops.py#L2550 assert that the input gradient must be nested (as the error suggests).

### Versions

<details>
<summary>Versions</summary>
<br>

On-line CPU(s) list:                  0-11
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) CPU @ 2.20GHz
CPU family:                           6
Model:                                85
Thread(s) per core:                   2
Core(s) per socket:                   6
Socket(s):                            1
Stepping:                             7
BogoMIPS:                             4400.53
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat avx512_vnni md_clear arch_capabilities
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            192 KiB (6 instances)
L1i cache:                            192 KiB (6 instances)
L2 cache:                             6 MiB (6 instances)
L3 cache:                             38.5 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-11
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Vulnerable
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:             Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Vulnerable; BHI: Vulnerable (Syscall hardening enabled)
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Vulnerable

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] nvtx==0.2.10
[pip3] optree==0.13.1
[pip3] pynvjitlink-cu12==0.4.0
[pip3] pytorch-triton==3.2.0+gitb2684bf3
[pip3] torch==2.7.0.dev20250204+cu124
[pip3] torchaudio==2.5.1+cu121
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.20.1+cu121
[conda] Could not collect
</details>

cc @cpuhrsch @jbschlosser @bhosmer @drisspg @soulitzer @davidberard98 @YuqingJ

### PR DESCRIPTION
**Pull request by michael-diggin**
Fixes #146404

Adds changes to the matmul and matmul_backward operation for nested jagged tensors, to support back propagation when the output is a regular strided tensor.
This required adding support for the nested matmul operation to work when the nested tensor wasn't 'self', i.e
`A@B` where `A` isn't nested but `B` is.

The operation schemas had to be updated to reflect that either input can be a strided tensor instead (and the gradient), so an extra assertion is added in an edge case where neither input is nested.

Unit tests are also added.

### CODE REVIEW

File: torch/nested/_internal/ops.py [hunk 1]
```diff
@@ -2559,6 +2587,14 @@ def matmul_backward_default(func, *args, **kwargs):
     other = new_kwargs.pop("other")
     grad_input_mask = new_kwargs.pop("mask")
 
+    if grad is None:
+        return (None, None)
+
+    if not inp.is_nested and not other.is_nested:
+        raise ValueError(
+            "matmul_backward_default: expected one of [self, other] to be a nested tensor"
+        )
```
Comment (soulitzer) [inline torch/nested/_internal/ops.py:2596]: You can't actually get to this `matmul_backward_default` function  if neither of the inputs are nested tensors

Comment (michael-diggin) [reply torch/nested/_internal/ops.py]: Yep true, removed - thanks!

File: test/test_nestedtensor.py [hunk 2]
```diff
@@ -2247,6 +2247,38 @@ def test_matmul_noncontiguous(self, device, dtype):
             torch.matmul(nt0_noncontiguous.transpose(-1, -2), nt1_noncontiguous),
         )
 
+    @dtypes(torch.float, torch.double)
+    def test_mamtul_dense_x_njt(self, device, dtype):
+        # T (B, D, E) with NJT (B, *, E, C) broadcasting case
```
Comment (soulitzer) [inline test/test_nestedtensor.py:2252]: random_nt_from_dims([3, None, 4]) produces (B, *, E) not (B, *, E, C) right

**Review by soulitzer — COMMENTED** [review]
Thanks for the PR! Added some small comments

Btw, there's some existing infra around testing that we can use that will automatically test forward, backward, compile, etc. (See the tests under TestNestedTensorOpInfo)

To test the new inputs that you are supporting here, instead of adding new one-off tests here, you can add the inputs that you would like to test to `sample_inputs_matmul` in `torch/testing/_internal/opinfo/definitions/nested.py`

File: torch/nested/_internal/ops.py [hunk 3]
```diff
@@ -1153,14 +1153,23 @@ def matmul_default(func, *args, **kwargs):
     inp = new_kwargs.pop("input")
     other = new_kwargs.pop("other")
 
+    if not inp.is_nested and not other.is_nested:
+        raise ValueError(
+            "matmul_default: expected one of [self, other] to be a nested tensor"
+        )
+
```
[the remark this answers is not in the document]
Comment (michael-diggin) [reply torch/nested/_internal/ops.py]: Removed

File: torch/nested/_internal/ops.py [hunk 4]
```diff
@@ -1202,6 +1215,21 @@ def _padded_impl(a, b):
...
+            # convert to padded for this
+            return _padded_impl(inp, other, False)
+        # Support broadcasting the dense:
+        # (D, E) x (B, E, j1) => (B, D, j1)
+        # (D, E) x (B, E, j1, F) => (B, D, j1, F)
+        # etc.
+        elif inp.dim() == 2 and other.dim() > inp.dim():
+            return NestedTensor(
+                func(inp, other._values, **new_kwargs), **extract_kwargs(other)
+            )
```
[the remark this answers is not in the document]
Comment (michael-diggin) [reply torch/nested/_internal/ops.py:1232]: Good call, added this to the various if clauses.

File: test/test_nestedtensor.py (hunk 2 above)
Comment (michael-diggin) [reply test/test_nestedtensor.py]: Yes, I had forgotten to update the comment, I've removed this test anyway now for the OpInfo one

**Comment by michael-diggin** [comment]
> Thanks for the PR! Added some small comments
> 
> Btw, there's some existing infra around testing that we can use that will automatically test forward, backward, compile, etc. (See the tests under TestNestedTensorOpInfo)
> 
> To test the new inputs that you are supporting here, instead of adding new one-off tests here, you can add the inputs that you would like to test to `sample_inputs_matmul` in `torch/testing/_internal/opinfo/definitions/nested.py`

Thanks for the review @soulitzer! I've made changes based on the comments, and updated the tests to use the existing test infra/setup which is much nicer.

File: torch/testing/_internal/opinfo/definitions/nested.py [hunk 5]
```diff
@@ -1021,6 +1021,53 @@ def sample_inputs_matmul(
...
+    for njt_3d in _sample_njts(
+        device=device,
+        dtype=dtype,
+        requires_grad=requires_grad,
+        dims=[3],
+    ):
+        # (B, E, j1) x (B, j1, F) => (B, E, F)
+        if njt_3d._ragged_idx == 2 and njt_3d.is_contiguous():
+            B, E, _ = njt_3d.shape
+            j1 = len(njt_3d.values())
```
Comment (soulitzer) [inline torch/testing/_internal/opinfo/definitions/nested.py:1057]:
```suggestion
            sum_j1 = len(njt_3d.values())
```
nit: Usually j1 refers to the ragged dimension, but what you are trying to express here is the entire length of the packed dimension, e.g. sum(j1)

**Review by soulitzer — APPROVED** [review]
Thanks for quick updates! Just have another nit

**Comment by michael-diggin** [comment]
> Thanks for quick updates! Just have another nit

Thanks for the quick review @soulitzer! Just committed that small change. Could you rerun the workflows when you get a chance? As I think that's needed before I can merge.

**Review by jbschlosser — APPROVED** [review]
Awesome, thanks for the PR! looks solid on my end, thanks for updating the generated sample inputs to cover this case :)
